File: src/lib/dropbox-service.ts
import { Dropbox, DropboxAuth } from 'dropbox';
import { getAutomergeBinary, mergeExternalBinary } from './automerge-storage';

const DROPBOX_FILE_PATH = '/myokr-data.automerge';

/**
 * Generates the PKCE OAuth 2.0 authorization URL.
 */
export async function getDropboxAuthUrl(clientId: string): Promise<{ url: string; codeVerifier: string }> {
  const auth = new DropboxAuth({ clientId });
  // Pass undefined for redirectUri to use the "no-redirect" copy/paste code flow
  // Use 'offline' to get a refresh token
  const url = await auth.getAuthenticationUrl(undefined as any, undefined, 'code', 'offline', undefined, 'none', true);
  const codeVerifier = auth.getCodeVerifier();
  return { url: url as string, codeVerifier };
}

/**
 * Exchanges the authorization code for a refresh token.
 */
export async function exchangeDropboxCode(clientId: string, code: string, codeVerifier: string): Promise<string> {
  const auth = new DropboxAuth({ clientId });
  auth.setCodeVerifier(codeVerifier);
  const response = await auth.getAccessTokenFromCode(undefined as any, code);
  return (response.result as any).refresh_token;
}

/**
 * Validates a Dropbox connection by calling the users/get_current_account API.
 */
export async function validateDropboxToken(clientId: string, refreshToken: string): Promise<boolean> {
  try {
    const dbx = new Dropbox({ clientId, refreshToken });
    await dbx.usersGetCurrentAccount();
    return true;
  } catch (error) {
    console.error('Dropbox validation failed', error);
    return false;
  }
}

/**
 * Downloads the automerge file from Dropbox.
 * Returns null if the file does not exist.
 */
export async function downloadFromDropbox(clientId: string, refreshToken: string): Promise<Uint8Array | null> {
  try {
    const dbx = new Dropbox({ clientId, refreshToken });
    const response = await dbx.filesDownload({ path: DROPBOX_FILE_PATH });
    
    // Dropbox API returns fileBlob in the response for browser environments
    const fileBlob = (response.result as any).fileBlob;
    if (fileBlob) {
      // Cap the remote blob size before materializing it: a pathological/huge doc
      // would otherwise be loaded+merged synchronously on the UI thread (Automerge
      // WASM), freezing or OOM-crashing the webview on every 5-min auto-sync.
      const MAX_SYNC_BYTES = 50 * 1024 * 1024; // 50 MB — well above any realistic OKR dataset
      if (fileBlob.size > MAX_SYNC_BYTES) {
        throw new Error(`Sync aborted: remote document (${fileBlob.size} bytes) exceeds the ${MAX_SYNC_BYTES} byte cap`);
      }
      const arrayBuffer = await fileBlob.arrayBuffer();
      return new Uint8Array(arrayBuffer);
    }
    return null;
  } catch (error: any) {
    // If the file is simply not found, return null
    if (error?.error?.error_summary?.includes('path/not_found')) {
      return null;
    }
    throw error;
  }
}

/**
 * Uploads the automerge binary file to Dropbox.
 */
export async function uploadToDropbox(clientId: string, refreshToken: string, binary: Uint8Array): Promise<void> {
  try {
    const dbx = new Dropbox({ clientId, refreshToken });
    await dbx.filesUpload({
      path: DROPBOX_FILE_PATH,
      contents: binary,
      mode: { '.tag': 'overwrite' },
    });
  } catch (error) {
    console.error('Dropbox upload failed', error);
    throw error;
  }
}

/**
 * Performs a full synchronization with Dropbox:
 * 1. Downloads the remote file
 * 2. Merges it with the local document
 * 3. Uploads the merged result
 * Returns true if successful and a merge happened, false if token invalid or no remote file.
 */
export async function syncWithDropbox(clientId: string, refreshToken: string): Promise<boolean> {
  if (!clientId || !refreshToken) return false;
  try {
    const remoteBinary = await downloadFromDropbox(clientId, refreshToken);
    let finalBinary: Uint8Array;
    
    if (remoteBinary) {
      finalBinary = await mergeExternalBinary(remoteBinary);
    } else {
      finalBinary = await getAutomergeBinary();
    }
    
    await uploadToDropbox(clientId, refreshToken, finalBinary);
    return true;
  } catch (error) {
    console.error('Sync failed:', error);
    throw error;
  }
}

File: mobile/lib/src/dropbox_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:http/http.dart' as http;

import 'package:crypto/crypto.dart';

class DropboxService {
  final http.Client _client;

  DropboxService({http.Client? client}) : _client = client ?? http.Client();

  /// Generates a PKCE code verifier and challenge.
  /// Returns a tuple: (codeVerifier, url)
  (String, String) getDropboxAuthUrl(String clientId) {
    // Generate code verifier
    final rand = Random.secure();
    final values = List<int>.generate(32, (i) => rand.nextInt(256));
    final codeVerifier = base64UrlEncode(values).replaceAll('=', '');

    // Generate code challenge
    final bytes = utf8.encode(codeVerifier);
    final digest = sha256.convert(bytes);
    final codeChallenge = base64UrlEncode(digest.bytes).replaceAll('=', '');

    // Construct URL
    final url = Uri.https('www.dropbox.com', '/oauth2/authorize', {
      'client_id': clientId,
      'response_type': 'code',
      'token_access_type': 'offline',
      'code_challenge': codeChallenge,
      'code_challenge_method': 'S256',
    }).toString();

    return (codeVerifier, url);
  }

  /// Exchanges the authorization code for a refresh token.
  Future<String> exchangeDropboxCode(String clientId, String code, String codeVerifier) async {
    final response = await _client.post(
      Uri.parse('https://api.dropboxapi.com/oauth2/token'),
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: {
        'grant_type': 'authorization_code',
        'code': code,
        'client_id': clientId,
        'code_verifier': codeVerifier,
      },
    );

    if (response.statusCode == 200) {
      final json = jsonDecode(response.body);
      return json['refresh_token'] as String;
    } else {
      throw Exception('Failed to exchange code: ${response.body}');
    }
  }

  /// Validates the token by getting current account
  Future<bool> validateDropboxToken(String clientId, String refreshToken) async {
    try {
      final accessToken = await _getAccessToken(clientId, refreshToken);
      final response = await _client.post(
        Uri.parse('https://api.dropboxapi.com/2/users/get_current_account'),
        headers: {
          'Authorization': 'Bearer $accessToken',
        },
      );
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  Future<String> _getAccessToken(String clientId, String refreshToken) async {
    final response = await _client.post(
      Uri.parse('https://api.dropboxapi.com/oauth2/token'),
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: {
        'grant_type': 'refresh_token',
        'refresh_token': refreshToken,
        'client_id': clientId,
      },
    ).timeout(const Duration(seconds: 10));

    if (response.statusCode == 200) {
      final json = jsonDecode(response.body);
      return json['access_token'] as String;
    } else {
      throw Exception('Failed to get access token: ${response.statusCode}');
    }
  }

  /// Downloads the automerge binary file from Dropbox.
  /// Returns null if the file does not exist.
  Future<Uint8List?> downloadFromDropbox(String clientId, String refreshToken) async {
    final accessToken = await _getAccessToken(clientId, refreshToken);
    final response = await _client.post(
      Uri.parse('https://content.dropboxapi.com/2/files/download'),
      headers: {
        'Authorization': 'Bearer $accessToken',
        'Dropbox-API-Arg': jsonEncode({'path': '/myokr-data.automerge'}),
      },
    ).timeout(const Duration(seconds: 10));

    if (response.statusCode == 200) {
      const maxSyncBytes = 50 * 1024 * 1024; // 50 MB cap
      if (response.bodyBytes.length > maxSyncBytes) {
        throw Exception('Sync aborted: remote document (${response.bodyBytes.length} bytes) exceeds 50MB cap');
      }
      return response.bodyBytes;
    } else if (response.statusCode == 409 || response.body.contains('path/not_found')) {
      return null;
    } else {
      throw Exception('Download failed with status ${response.statusCode}: ${response.body}');
    }
  }

  /// Uploads the automerge binary file to Dropbox.
  Future<void> uploadToDropbox(String clientId, String refreshToken, List<int> binary) async {
    final accessToken = await _getAccessToken(clientId, refreshToken);
    final response = await _client.post(
      Uri.parse('https://content.dropboxapi.com/2/files/upload'),
      headers: {
        'Authorization': 'Bearer $accessToken',
        'Content-Type': 'application/octet-stream',
        'Dropbox-API-Arg': jsonEncode({
          'path': '/myokr-data.automerge',
          'mode': 'overwrite',
          'autorename': false,
          'mute': false,
          'strict_conflict': false,
        }),
      },
      body: binary,
    ).timeout(const Duration(seconds: 10));

    if (response.statusCode != 200) {
      throw Exception('Upload failed with status ${response.statusCode}: ${response.body}');
    }
  }


  /// Performs full synchronization with Dropbox:
  /// 1. Downloads remote file (if forceUpload/compacted is false)
  /// 2. Merges with local binary callback
  /// 3. Uploads merged result
  Future<bool> syncWithDropbox({
    required String clientId,
    required String refreshToken,
    required Future<Uint8List> Function() getLocalBinary,
    required Future<Uint8List> Function(Uint8List remoteBinary) mergeExternalBinary,
    bool forceUpload = false,
    bool compactedSinceLastSync = false,
  }) async {
    if (clientId.isEmpty || refreshToken.isEmpty) return false;

    Uint8List finalBinary = await getLocalBinary();

    if (!forceUpload && !compactedSinceLastSync) {
      final remoteBinary = await downloadFromDropbox(clientId, refreshToken);
      if (remoteBinary != null && remoteBinary.isNotEmpty) {
        finalBinary = await mergeExternalBinary(remoteBinary);
      }
    }

    await uploadToDropbox(clientId, refreshToken, finalBinary);
    return true;
  }
}

